### PR TITLE
A few fixes for compiling with nvc++

### PR DESCRIPTION
**Pull request by bernhardmgruber**
# This Pull request:

## Changes or fixes:

Fixes several issues when compiling ROOT with nvc++, but not all of them. The build still fails ultimately, because `clang-tblgen` crashes. See also #9036.

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

### CODE REVIEW

**Comment by bellenot** [comment]
@phsft-bot build

File: interpreter/llvm-project/llvm/include/llvm/ADT/PointerUnion.h [hunk 1]
```diff
@@ -261,8 +261,9 @@ struct PointerLikeTypeTraits<PointerUnion<PTs...>> {
 
   // The number of bits available are the min of the pointer types minus the
   // bits needed for the discriminator.
-  static constexpr int NumLowBitsAvailable = PointerLikeTypeTraits<decltype(
-      PointerUnion<PTs...>::Val)>::NumLowBitsAvailable;
+  using nvhpcWorkAround = decltype(PointerUnion<PTs...>::Val);
```
Comment (vgvassilev) [inline interpreter/llvm-project/llvm/include/llvm/ADT/PointerUnion.h:264]: Is this patch a backport of the llvm upstream?

Comment (bernhardmgruber) [reply interpreter/llvm-project/llvm/include/llvm/ADT/PointerUnion.h:264]: No, LLVM upstream (17.0.6) does not have this change and nvc++ fails equally there. What is the procedure to apply such a workaround to ROOT's LLVM?

(while NVIDIA may be working on a fix to nvc++)

Comment (vgvassilev) [reply interpreter/llvm-project/llvm/include/llvm/ADT/PointerUnion.h:264]: I think it should be enough if we open a PR against LLVM mainline and backport it here. It'd be better to wait until it gets merged upstream but generally there is no need to.

cc: @hahnjo

Comment (hahnjo) [reply interpreter/llvm-project/llvm/include/llvm/ADT/PointerUnion.h:264]: Either way, I'm not a fan of introducing a workaround to our copy of LLVM for a very specific compiler. I believe upstream has a very clear documentation on which compilers it supports, and NVIDIA's isn't one of them. It's not our job to fix this downstream.

**Comment by bernhardmgruber** [comment]
Alright, this gets more tricky than anticipated. I had to workaround a few more codegen/optimizer bugs in nvc++. I changed the PR to draft for now.

**Comment by hahnjo** [comment]
I propose to close this: As I expressed above, I believe it's not our job to fix upstream LLVM to compile with `nvc++`.

**Comment by bernhardmgruber** [comment]
Fine for me, I will open a separate PR with the reasonable changes.

**Pull request closed without merging**